### PR TITLE
Tickets/dm 17812

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: www/qserv/js/ToolsSql.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+            let html = `
+<div id="fwk-tools-sql">
+  <div class="row">
+    <div class="col">
+      <p>This tool allows executing queries directly against the <span style="font-weight:bold;">MariaDB</span>
+        service of a <span style="font-weight:bold;">Qserv</span> worker. For queries which produce a result set
+        the result will be displayed as a table below the form. Errors will be also reported.
+        Please, <span style="font-weight:bold;">DO NOT</span> launch queries which may produce millions of rows.
+        An attept to do so will make thsi applicatio unresponsive.
```
Comment (andy-slac) [inline www/qserv/js/ToolsSql.js]: attept thsi applicatio 

It already feels half-unresponsive, some characters are lost 😉

File: www/qserv/js/ToolsSql.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        Use <span style="font-style:italic;">SELECT COUNT(*)</span> first.
```
Comment (andy-slac) [inline www/qserv/js/ToolsSql.js]: Usually people use monospace for pieces of code, not italic.

File: www/qserv/js/ToolsSql.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                this._password_obj = this.fwk_app_container.find('input#password');
+            }
+            return this._password_obj;
+        }
+        _query() {
+            if (this._query_obj === undefined) {
+                this._query_obj = this.fwk_app_container.find('textarea#query');
+            }
+            return this._query_obj;
+        }
```
Comment (andy-slac) [inline www/qserv/js/ToolsSql.js]: I see a lot of repetition in these methods, maybe a chance to refactor a bit?

File: www/qserv/js/ToolsSql.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+        /**
+         * Load data from a web servie then render it to the application's page.
+         */
+        _load() {
+            if (this._loading === undefined) {
+                this._loading = false;
+            }
+            if (this._loading) return;
+            this._loading = true;
```
Comment (andy-slac) [inline www/qserv/js/ToolsSql.js]: Isn't there a race condition here? Does JS have mutexes to protect it? Does this question even make sense? 😈

File: www/qserv/js/ToolsSql.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                }
+            );
+        }
+        
+        /**
+         * Analyze and render result sets or error
+         */
+        _display(data) {
+            console.log(data);
+            if (data.status !== 'FINISHED::SUCCESS::EXT_STATUS_NONE') {
```
Comment (andy-slac) [inline www/qserv/js/ToolsSql.js]: Magic!

File: www/qserv/js/ToolsSql.js (hunk 4 above)
Comment (iagaponenko) [reply www/qserv/js/ToolsSql.js]: JS is:
- single-threaded
- event-driven
Hence, no synchronization is needed. Each event handler executes within a safe environment. In fact, every code sequence  in JS is an "event". Events are serialized by a browser.

File: www/qserv/js/ToolsSql.js (hunk 5 above)
Comment (iagaponenko) [reply www/qserv/js/ToolsSql.js]: Good point. I could add a boolean flag to indicate if a result of an operation was successful.

File: www/qserv/js/ToolsSql.js (hunk 3 above)
Comment (iagaponenko) [reply www/qserv/js/ToolsSql.js]: Good point. I will have a look at it.

File: www/qserv/js/ToolsSql.js (hunk 2 above)
Comment (iagaponenko) [reply www/qserv/js/ToolsSql.js]: Good suggestion!

File: www/qserv/js/ToolsSql.js (hunk 1 above)
Comment (iagaponenko) [reply www/qserv/js/ToolsSql.js]: Noted! Will fix it.

**Pull request merged**